source/_posts/2025-05-12-mcp-explained-without-hype-or-fluff.markdown

```diff
@@ -0,0 +1,160 @@
+
+One way for me to have approached this is to write an AI application from scratch encoded with knowledge about all the CKAN REST APIs. Unfortunately, this would have "locked in" AI use of CKAN open data sets to just my application. [And data, especially Open Data wants to be free](https://en.wikipedia.org/wiki/Information_wants_to_be_free).
+
+What I really wanted is a well-known "doorknob" that a lot of AI applications and agents in the world would know how to open. This is what MCP servers do. I wrote one in a couple of hours.
+
+I used the official MCP Python SDK and defined some tools. Here's an excerpt of what that looks like:
+
+```python
+@mcp.tool()
+async def list_tags(query: Optional[str] = None, limit: int = 50, ctx: Context = None) -> str:
+    """List available tags in CKAN.
+
+    Args:
+        query: Optional search string to filter tags
+        limit: Maximum number of tags to return
+
+    Returns:
+        A formatted string containing available tags.
+    """
+    # code to list all the tags used to tag data, via the CKAN API
+
+@mcp.tool()
+async def search_datasets(
+    query: Optional[str] = None,
+    tags: Optional[List[str]] = None,
+    organization: Optional[str] = None,
+    format: Optional[str] = None,
+    limit: int = 10,
+    offset: int = 0,
+    ctx: Context = None
+) -> str:
+    """Search for datasets with various filters.
+
+    Args:
+        query: Free text search query
+        tags: Filter by tags (list of tag names)
+        organization: Filter by organization name
+        format: Filter by resource format (e.g., CSV, JSON)
+        limit: Maximum number of datasets to return
+        offset: Number of datasets to skip
+
+    Returns:
+        A formatted string containing matching datasets.
+    """
+    # code to handle searches, using the CKAN API
+
+@mcp.tool()
+async def get_resource_details(resource_id: str, ctx: Context = None) -> str:
+    """Get detailed information about a specific resource (file/data).
+
+    Args:
+        resource_id: The ID of the resource
+
+    Returns:
+        A formatted string containing resource details.
+    """
+    # code to read the details and get the link to a specific resource, using the CKAN API
+```
+
+The details of the SDK are better explained in official guides, but the gist of it is that it is an abstraction over JSON-RPC request-response messages that are defined in the protocol. The server I have implemented runs locally, launched as a subprocess by the client app and uses the stdio streams to pass these protocol messages around. Remote MCP servers are a thing as well.
+
+After I wrote this server, I exposed it to the Claude desktop app, which is also an MCP client by editing `claude_desktop_config.json`. I pointed it to [JusticeHub](https://justicehub.in), a CKAN instance that contains legal and justice data, created by the folks at [CivicDataLabs](https://civicdatalab.in/).
```

Comment:
It might be useful to talk about how you came across JusticeHub and also mention that finding datasets is also a thing you could do with this MCP server